### ISSUE TITLE
Licensing Question regarding libiberty (LGPL)

### ISSUE THREAD

**Issue by benstadin** [facebook/folly#1466]
I'm checking the license of folly, including depending licenses. Folly itself is Apache 2, but it includes libiberty which is LGPL. This can lead to issues in certain situations, I therefore want to clarify whether building Folly with thde default build options using cmake also links to libiberty. 

This is not quite clear to me. Libiberty is looked up in the cmake file, but I'm ot quite sure whether FOLLY_DETAIL_HAVE_DEMANGLE_H is set by default or the no-op implementation is applied in Folly Demangle.

**Comment by benstadin** [comment]
Could this be delegated to a responsible person?

**Comment by Orvid** [comment]
If demangle.h is found, folly will attempt to link against libiberty, which, depending on the system configuration, typically means it links against the shared library version of libiberty installed on the system.

**Comment by benstadin** [comment]
However, since this opens kind of backdoor licensing issue, it would be better that the default behaviour would be to neither include nor link against libiberty. 

Alternatively (guessing the first proposal would break backward compatibility at build time) it would be nice to have a cmake option like NO_LIBIBERTY to avoud including demabgle.

**Comment by benstadin** [comment]
It's however better though to have certainty, and control over this. E.g. in case where folly is accidentally linked against it statically.

**Comment by daverigby** [comment]
Given you have control over your build environment, why don’t you just ensure libiberty isn’t installed? Then you’re golden :)

**Comment by emmenlau** [comment]
@daverigby I guess not everyone is working only on their self-administrated machines (think about Universities, larger companies, etc)

**Comment by daverigby** [comment]
@emmenlau True, however I'd argue if you were in such an environment and LGPL code was problematic to you, it probably shouldn't be installed (by whoever _does_ admin the machine) in the first place :)

**Comment by emmenlau** [comment]
@daverigby Is there any reason as to why you would think that?

**Comment by daverigby** [comment]
I'm somewhat playing devil's advocate (given I think the question above has been answered - libiberty _will_ be used if present and detected by CMake) - but if an organisation has an issue with linking against LGPL'd libraries for programs they create, then it seems prudent to _not_ make such libraries available - by virtue of not installing the *-dev package for example.

**Comment by emmenlau** [comment]
I can see how its fun to play devils advocate. But at the end of the day its not really helping. We should maybe cut this there.

**So the problem stands**, currently folly has no option to choose linking against `libiberty`, and possibly the available `libiberty` on the system may be a static library, which in turn becomes a cmake transitive dependency of folly, possibly tainting any executable with an LGPL license.

I can see that this can be a problem for people, and it is a problem for me (albeit I did not check if any system I'm using ships a static `libiberty`.

**Comment by daverigby** [comment]
BTW there's other LGPL libraries potentially linked - for example libaio - see https://github.com/facebook/folly/blob/505affdcff9184b593627840e06fca01134086b7/CMake/folly-deps.cmake#L136

You might want to check if that (and possibly) others are a problem for you.

**Comment by benstadin** [comment]
We now maintain a docker build environment and a script in our company to ensure that no GPL licensed library is installed on the system at compile time of folly. 

As much as I praise Facebook for open sourcing the Folly library, I still need to emphasize that this needs to be addressed. If the library is licensed under Apache 2, then you (facebook) have to provide the means to ensure it IS licensable under Apache 2. In praxis. Not just in theory, requiring any user of the library to deep dive it’s dependencies and take the risk of trapping into a licensing problem with his own product. 

This is a serious problem.
